src/common/io/tls/client.c

```diff
@@ -0,0 +1,544 @@
+                    "unable to set user-defined CA certificate location");
+            }
+            // Else use the defaults
+            else
+                cryptoError(SSL_CTX_set_default_verify_paths(this->context) != 1, "unable to set default CA certificate location");
+        }
+    }
+    MEM_CONTEXT_NEW_END();
+
+    FUNCTION_DEBUG_RESULT(TLS_CLIENT, this);
+}
+
+/***********************************************************************************************************************************
+Convert an ASN1 string used in certificates to a String
+***********************************************************************************************************************************/
+static String *
+asn1ToStr(ASN1_STRING *nameAsn1)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(VOIDP, nameAsn1);
+    FUNCTION_TEST_END();
+
+    // The name should not be null
+    if (nameAsn1 == NULL)
+        THROW(CryptoError, "TLS certificate name entry is missing");
+
+    FUNCTION_TEST_RESULT(
+        STRING, strNewN(
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+            (const char *)ASN1_STRING_data(nameAsn1),
+#else
+            (const char *)ASN1_STRING_get0_data(nameAsn1),
+#endif
+            (size_t)ASN1_STRING_length(nameAsn1)));
+}
+
+/***********************************************************************************************************************************
+Check if a name from the server certificate matches the hostname
+
+Matching is always case-insensitive since DNS is case insensitive.
+***********************************************************************************************************************************/
+static bool
+tlsClientHostVerifyName(const String *host, const String *name)
+{
+    FUNCTION_DEBUG_BEGIN(logLevelTrace);
+        FUNCTION_DEBUG_PARAM(STRING, host);
+        FUNCTION_DEBUG_PARAM(STRING, name);
+
+        FUNCTION_DEBUG_ASSERT(host != NULL);
+        FUNCTION_DEBUG_ASSERT(name != NULL);
+    FUNCTION_DEBUG_END();
+
+    // Reject embedded nulls in certificate common or alternative name to prevent attacks like CVE-2009-4034
+    if (strlen(strPtr(name)) != strSize(name))
+        THROW(CryptoError, "TLS certificate name contains embedded null");
+
+    bool result = false;
+
+    // Try an exact match
+    if (strcasecmp(strPtr(name), strPtr(host)) == 0)
+    {
+        result = true;
+    }
+    // Else check if a wildcard certificate matches the host name
+    //
+    // The rules are:
+    // 1. Only match the '*' character as wildcard
+    // 2. Only match wildcards at the start of the string
+    // 3. The '*' character does *not* match '.', meaning that we match only a single pathname component.
+    // 4. Don't support more than one '*' in a single pattern.
+    //
+    // This is roughly in line with RFC2818, but contrary to what most browsers appear to be implementing (point 3 being the
+    // difference)
```

Comment:
I think 3. does not allow for Subject Alternative names - and maybe we should? (https://en.wikipedia.org/wiki/Wildcard_certificate#Example)

Reply:
I see you do mention SAN support below, so my only comment is if this means the "naked" domain should be considered valid.

Reply:
The naked domain is valid if there is an exact match for it.